http-clients/src/main/java/org/triplea/http/client/web/socket/WebsocketListener.java

```diff
@@ -0,0 +1,100 @@
+package org.triplea.http.client.web.socket;
+
+import com.google.common.base.Preconditions;
+import java.net.URI;
+import java.util.Optional;
+import java.util.function.Consumer;
+import lombok.AccessLevel;
+import lombok.Getter;
+import org.triplea.http.client.web.socket.messages.ServerMessageEnvelope;
+import org.triplea.http.client.web.socket.messages.WebsocketMessageType;
+
+/**
+ * Used to route server websocket messages to a listener.
+ *
+ * <p>Caution: Do not forget to call the {@code close} method when done listening!
+ *
+ * <p>Very-generic websocket listener! Knows about a set of listeners and message types. The message
+ * types know how to extract a specific message from a server message payload and to send that
+ * message to a listener from the set of listeners. This class wires those together, the net effect
+ * is that after construction a websocket connection is created and when messages are received they
+ * will be routed to the correct listener.
+ *
+ * @param <MessageTypeT> Websocket message type enum.
+ * @param <ListenersTypeT> Listener class, holds a set of consumers, each is an individual consumer.
+ */
+public abstract class WebsocketListener<
+        MessageTypeT extends WebsocketMessageType<ListenersTypeT>, ListenersTypeT>
+    implements Consumer<ServerMessageEnvelope> {
+
+  @Getter(AccessLevel.PROTECTED)
+  private final GenericWebSocketClient webSocketClient;
+
+  private ListenersTypeT listeners;
+
+  private WebsocketListener(
+      final GenericWebSocketClient genericWebSocketClient, final ListenersTypeT listeners) {
+    webSocketClient = genericWebSocketClient;
+    webSocketClient.addMessageListener(this);
+    this.listeners = listeners;
+  }
+
+  /**
+   * Use this constructor if the listener will also be an outbound message sender. Injecting the
+   * websocket client directly allows for easier testing & mocking.
+   */
+  protected WebsocketListener(final GenericWebSocketClient genericWebSocketClient) {
+    this(genericWebSocketClient, null);
+  }
+
+  /**
+   * Constructs a websocket listener that will be connected and messages will be routed to the
+   * listeners passed in.
+   *
+   * @param hostUri Server host (base) uri.
+   * @param websocketPath Path on the server to the websocket to which we will listen.
+   * @param listeners Listener object that contains listeners for each message type we would expect
+   *     to receive from server.
+   */
+  protected WebsocketListener(
+      final URI hostUri, final String websocketPath, final ListenersTypeT listeners) {
+    this(
+        new GenericWebSocketClient(
+            URI.create(hostUri + websocketPath),
+            "Failed to connect to " + URI.create(hostUri + websocketPath)),
```

Comment:
Hmm is it really necessary to create a URI object for the error message?
If `URI.create("http://example.com").toString()` really returns something different than `"http://example.com" then I would consider putting this error message inside the `GenericWebSocketClient` constructor, or accept a lambda that maps a URI object to a message so that the object doesn't have to be created 2 times.

Reply:
I'm starting to forget why the error message is parameterized. I'd like to personally see it be updated so that 'new' of the websocket does not create a connection and that be done explicitly later. It's been a challenge in test to work with that. It's a balance though as we would then need to be sure to call a 'connect' function. 

At the moment I don't know if it's clear that any 'genericWebSocketClient' will always want the same connection error message. Because we have multiple websockets now, knowing the path to the websocket seems handy. You did not that `hostUri` is of type URI? So we're creating the fully qualified path, because we're calling another constructor we are pretty forced to do the create twice. It's a little ugly, IMO ultimately the technical debt for it is extremely localized and not impactful. It seemed like then a decent option compared to going to other lengths.

Reply:
Well I don't think accepting a `Function<URI, String>` isn't a lot more complicated than accepting a plain string, but on the other hand you could create a private static helper method that creates a `GenericWebSocketClient` for you and re-uses the URI object.

Reply:
> Well I don't think accepting a Function<URI, String> isn't a lot more complicated 

That could be done, but conceptually it's somewhat difficult. We pass a URI to a super class, where we also pass a function, where we assume the input to that function will be the same URI that we passed to the super class. I don't think client classes should be quite that aware of the super-class implementation.

I don't know if DRY'ing up the URI creation is the right thing to do, it could be the wrong fix. I'm concerned about code churn in this area. I'd prefer to leave a code duplication code smell in favor of seeing how the usages develop as we add more examples.